collector/exporter/otelarrowexporter/internal/arrow/stream.go

```diff
@@ -227,22 +229,25 @@ func (s *Stream) run(bgctx context.Context, streamClient StreamClientFunc, grpcO
 		s.logStreamError("writer", writeErr)
 	}
 
+	s.workState.lock.Lock()
+	defer s.workState.lock.Unlock()
+
 	// The reader and writer have both finished; respond to any
 	// outstanding waiters.
-	for _, ch := range s.waiters {
+	for _, ch := range s.workState.waiters {
 		// Note: the top-level OTLP exporter will retry.
 		ch <- ErrStreamRestarting
 	}
+
+	s.workState.waiters = map[int64]chan<- error{}
 }
 
 // write repeatedly places this stream into the next-available queue, then
 // performs a blocking send().  This returns when the data is in the write buffer,
 // the caller waiting on its error channel.
-func (s *Stream) write(ctx context.Context) (retErr error) {
+func (s *Stream) write(ctx context.Context, cancel context.CancelFunc) (retErr error) {
```

Comment:
Is the `context.CancelFunc` needed in this function?

Reply:
Fixed.